cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java

```diff
@@ -157,8 +163,23 @@ public Long apply(StreamProperty property) {
 
   @Override
   public Cancellable addListener(String streamName, StreamPropertyListener listener) {
-    return propertyStore.get().addChangeListener(streamName, new StreamPropertyChangeListener(streamAdmin,
-                                                                                              streamName, listener));
+    return propertyStore.get().addChangeListener(streamName,
+                                                 new StreamPropertyChangeListener(streamAdmin, streamName, listener));
+  }
+
+  @Override
+  public Cancellable addLeaderCallback(final StreamLeaderCallback callback) {
+    synchronized (this) {
+      leaderCallbacks.add(callback);
```

Comment:
Better not to add a user object to a set (as you don't know what contract they've implemented for equals/hashCode) for the canceling/removal purpose. It's better be a wrapped one.